### DISCUSSION TITLE
How to use "reset" correctly?

### DISCUSSION THREAD

**Discussion by ZhyGao** [genn-team/genn#535] [Q&A]
In GeNN, I stitched together two identical 1650ms input data to generate a 3300ms input. I'm trying to use the reinitialise() command at 1650ms to reset the state of the network to run the second piece of data in the same connection mode. But calculating the instantaneous firing rate of the spikes after I simulated the network gave me the following graph. The firing rate simulated in the first segment of the data is completely different from that in the second segment. The connection between the populations in my model is using the "FixedNumberTotalWithReplacement" mode, we want to run the two pieces of data with the same connection matrix, can someone please help me?

![bdccbb1e57e6c5f403ce3f2f2b2f77e](https://user-images.githubusercontent.com/39367799/178138143-f24d312f-ba1e-4f05-8f17-d8847c4792de.png)

**Comment by neworderofjamie** [comment]
How are you presenting the data to the network?

**Reply by ZhyGao** [reply]
We want to use GeNN to build an SNN based liquid state machine and use it to perform classification tasks on the IBM Gesture dataset. How can I achieve such a task without changing the network connection mode? I'm currently using the method you mentioned in https://github.com/genn-team/genn/discussions/532. Thanks！

**Comment by neworderofjamie** [comment]
what I meant was, how are you connecting the gesture dataset to the model? Are you using the built in spike source array model?

**Reply by ZhyGao** [reply]
Yes, I use it. Each piece of data in the gesture dataset is expanded into 128x128x2 points, and each point is represented by a neuron. And connect the gesture dataset to the model by when each neuron fires.

**Comment by neworderofjamie** [comment]
How are you resetting the ``startSpike`` and ``endSpike`` state variables; and the ``spikeTimes`` extra global parameters? With ``spikeTimes`` I would recommend initialising it with an empty array large enough for the stimuli with the largest number of spikes and copying stimuli into that. Additionally, ``reinitialize`` doesn't reset the simulation time so you might want to do:
```
model.t = 0.0
model.timestep = 0
```
as well if your stimuli all start from t = 0.

**Reply by ZhyGao** [reply]
I spliced each Gesture dataset data stream to be input into the network into one, and then reset the network state at certain time intervals. In addition, I also tried another method, which is to use a spikeTrain with a very large number of spikes to build the model, reset the model state after simulating for a period of time, and input a new spikeTrain, specifically:spike_counts = [len(n) for n in spike_times ]

end_spike = np.cumsum(spike_counts)
start_spike = np.empty_like(end_spike)
start_spike[0] = 0
start_spike[1:] = end_spike[0:-1]

....

stim = model.add_neuron_population("Stim", nuymber, "SpikeSourceArray", {}, 
                                  {"startSpike": start_spike, "endSpike": end_spike})

model.add_synapse_population(...)

model.build()
model.load()
if model.t % Time == 0.:
    model.reinitialise()
    stim.set_extra_global_param("spikeTimes",  np.concatenate(spike_times))
model.step_time()

But an error was reported in model.step_time(): 
RuntimeError: runner.cc: 2174: cuda error 700 : an illegal memory access was encountered

Do you know how to accomplish such an input and network state reset? Thanks!

**Comment by neworderofjamie** [comment]
I would recommend doing something like this. Firstly pre-process the spike times associated with each stimuli into GeNN format and calculate the maximum number of spikes across all stimuli:
```python
stim_spikes = []
max_spikes = 0
for spike_times in stim_spike_times:
    spike_counts = [len(n) for n in spike_times]
    end_spike = np.cumsum(spike_counts)

    stim_spikes.append((spike_times, end_spike))
    max_spike_times = max(max_spike_times, len(spike_times))
```

Then setup the spike source array, reserving enough memory for the **maximum** number of spikes:
```python
stim = model.add_neuron_population("Stim", number, "SpikeSourceArray", {},
                                   {"startSpike": 0, "endSpike": 0})

stim.set_extra_global_param("spikeTimes", np.empty(max_spike_time))
```

Build and load the model:
```python
model.build()
model.load()
```

Loop through the pre-processed spike times and present each one to the network:
```python
for spike_times, end_spikes in stim_spikes:
    # Reinitialize model
    model.reinitialise()

    # Calculate start spikes
    start_spike[0] = 0
    start_spike[1:] = end_spike[0:-1]
    
    # Copy start and end spike indices into view
    stim.vars["startSpike"].view[:] = start_spike
    stim.vars["endSpike"].view[:] = end_spike

    # Copy spike times into (start of) view
    stim.extra_global_params["spikeTimes"][:len(spike_times)] = spike_times
    
    # Copy
    stim.push_var_to_device("startSpike")
    stim.push_var_to_device("endSpike")
    stim.push_extra_global_param_to_device("endSpike")

    # **TODO** simulate
```
